core/src/main/scala/kafka/server/BrokerToControllerChannelManager.scala

```diff
@@ -254,7 +256,7 @@ class BrokerToControllerChannelManagerImpl(
   def controllerApiVersions(): Option[NodeApiVersions] = {
     requestThread.activeControllerAddress().flatMap { activeController =>
       if (activeController.id == config.brokerId)
-        Some(currentNodeApiVersions)
+        currentNodeControllerApiVersions
```

Comment:
I wonder if this optimization is worth preserving. Any harm using the same negotiation even when the controller is co-resident?

Reply:
Yes, there may be a similar problem with finalized features since they will also change, I will check `KafkaApis.handleInvalidVersionsDuringForwarding` to inspect it.